modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk.clj

```diff
@@ -146,6 +154,90 @@
+                             int-fingerprint-types (fn [^String v]
+                                                     (java.math.BigInteger. v)) ; convert to BigInteger
+                             dec-fingerprint-types (fn [^String v]
+                                                     (java.math.BigDecimal. v)) ; convert to BigDecimal
+                             identity))) ; otherwise, return value unchanged
+          (sorted-map)
+          fields))
+
+(defn- FieldValueList->fingerprint-structure
+  "For the given `values` (of type [[FieldValueList]]), parse each value using the respective parser function from
+  `indexed-parser-fns` (expected to have been returned by [[fields->indexed-parser-fns->parser-fns]])."
+  [indexed-parser-fns ^FieldValueList values]
+  (reduce-kv (fn [acc ^Integer idx parser-fn]
+               (let [^FieldValue fv (.get values idx)
+                     v              (.getValue fv)
+                     parsed         (if (string? v)
+                                      (parser-fn v)
+                                      (str v))]
+                 (conj acc parsed)))
+             []
+             indexed-parser-fns))
+
+;; Override the table-rows-sample multimethod for :bigquery-cloud-sdk, to use a different mechanism for sampling
+;; table rows (the list operations, rather than a regular query), purely as a cost control measure.  For certain types
+;; of tables, BigQuery evidently does not honor the limit parameter of a query, and instead performs a full table scan
+;; and then truncates results at the end.  See #19470 for full details.
+(defmethod @#'sync.f/table-rows-sample :bigquery-cloud-sdk [driver
+                                                            {table-name :name, dataset-id :schema :as table}
+                                                            fields
+                                                            rff
+                                                            opts]
+  (let [database (table/database table)
+        bq-tbl   (get-table database dataset-id table-name)
+        tbl-type (-> (.getDefinition bq-tbl)
+                     (.getType))]
+    (if (contains? #{TableDefinition$Type/VIEW TableDefinition$Type/MATERIALIZED_VIEW} tbl-type)
+      (do (log/debugf "%s.%s is a view, so we cannot use the list API; falling back to regular query"
+                      dataset-id
+                      table-name)
+          ((get-method @#'sync.f/table-rows-sample :sql-jdbc) driver table fields rff opts))
+      (let [^TableResult rows  (.list bq-tbl (u/varargs BigQuery$TableDataListOption))
+            num-rows           metadata-queries/max-sample-rows
+            indexed-parser-fns (fields->indexed-parser-fns fields)]
+        (transduce (comp (take num-rows)
+                         (map (partial FieldValueList->fingerprint-structure indexed-parser-fns)))
+                   (rff nil) ; nil param is for JDBC ResultSet metadata, which doesn't apply here
```

Comment:
This isn't 100% true. The `metadata` param is supposed to be for **initial** metadata which is usually just column name a base type/database type. It's not specific to JDBC since other drivers like Mongo would have that metadata too.

It does turn out however that `metadata` actually isn't used -- https://github.com/metabase/metabase/blob/master/src/metabase/sync/analyze/fingerprint.clj#L48 -- so it doesn't make a difference here.